pkg/controller/master/backup/backup_target.go

```diff
@@ -79,32 +80,79 @@ func (h *TargetHandler) OnBackupTargetChange(key string, setting *harvesterv1.Se
 		return setting, err
 	}
 
-	// Since we remove S3 access key id and secret access key after S3 backup target has been verified,
-	// we stop the controller to reconcile it again.
-	if target.Type == settings.S3BackupType && (target.SecretAccessKey == "" || target.AccessKeyID == "") {
-		return nil, nil
-	}
+	logrus.Debugf("backup target change:%s:%s", target.Type, target.Endpoint)
 
-	if err = h.updateLonghornTarget(target); err != nil {
-		return nil, err
-	}
+	switch target.Type {
+	case settings.S3BackupType:
+		// Since S3 access key id and secret access key are stripped after S3 backup target has been verified
+		// in reUpdateBackupTargetSettingSecret
+		// stop the controller to reconcile it again.
+		if target.SecretAccessKey == "" && target.AccessKeyID == "" {
+			return nil, nil
+		}
+
+		if err = h.updateLonghornTarget(target); err != nil {
+			return nil, err
+		}
 
-	if target.Type == settings.S3BackupType {
 		if err = h.updateBackupTargetSecret(target); err != nil {
 			return nil, err
 		}
-	}
 
-	return h.updateBackupTargetSetting(setting.DeepCopy(), target, err)
+		return h.reUpdateBackupTargetSettingSecret(setting, target, err)
+
+	case settings.NFSBackupType:
+		if err = h.updateLonghornTarget(target); err != nil {
+			return nil, err
+		}
+
+		// delete the may existing previous secret of S3
+		if err = h.deleteBackupTargetSecret(target); err != nil {
+			return nil, err
+		}
+
+		// same operation as S3 does, but it looks to be not as expected
+		// the framework will set the True/False according to the final return of here
+		harvesterv1.SettingConfigured.SetError(setting, "", err)
+		return nil, nil
+
+	default:
+		// reset backup target to default, then delete related settings
+		if target.IsDefaultBackupTarget() {
+			// default backup target is not usable, longhorn may not support default (blank) backup target
+			// delete it
+			if err = h.deleteLonghornTarget(target); err != nil {
+				return nil, err
+			}
+
+			// delete the may existing previous secret of S3
+			if err = h.deleteBackupTargetSecret(target); err != nil {
+				return nil, err
+			}
+
+			// it has no effect to call
+			//  harvesterv1.SettingConfigured.SetError(setting, "", fmt.Errorf("reset value to default"))
+			//  to set True/False of a config, as the operation itself will finally return nil, nil
+			// hence
+			//  it is not effective to call harvesterv1.SettingConfigured.IsTrue(targetSetting), e.g. in checkBackupTargetConfigured
+			//  to check if the backup target is set or not
+			//  as the "use the default value" action itself will be success, it "reset" the value to default, but itself is successful
+			// the explicit check of "default value" is added to related parts
+			return nil, nil
```

Comment:
an empty string is allowed for the LH backup target, so don't delete the setting CR.
also can we simply the return comments, thanks.

Reply:
updated